node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,684 @@
+	/// Our current view.
+	view: View,
+
+	/// Additional data particular to a relay parent.
+	per_relay_parent: HashMap<Hash, PerRelayParentData>,
+}
+
+/// Data for a particular relay parent.
+#[derive(Debug, Clone, Default)]
+struct PerRelayParentData {
+	/// Signing context for a particular relay parent.
+	signing_context: SigningContext<Hash>,
+
+	/// Set of validators for a particular relay parent.
+	validator_set: Vec<ValidatorId>,
+
+	/// Set of validators for a particular relay parent for which we
+	/// received a valid `BitfieldGossipMessage`.
+	/// Also serves as the list of known messages for peers connecting
+	/// after bitfield gossips were already received.
+	one_per_validator: HashMap<ValidatorId, BitfieldGossipMessage>,
+
+	/// which messages of which validators were already sent
+	message_sent_to_peer: HashMap<PeerId, HashSet<ValidatorId>>,
+}
+
+impl PerRelayParentData {
+	/// Determines if that particular message signed by a validator is needed by the given peer.
+	fn message_from_validator_needed_by_peer(
+		&self,
+		peer: &PeerId,
+		validator: &ValidatorId,
+	) -> bool {
+		if let Some(set) = self.message_sent_to_peer.get(peer) {
+			!set.contains(validator)
+		} else {
+			false
+		}
+	}
+}
+
+fn network_update_message(n: NetworkBridgeEvent) -> AllMessages {
+	AllMessages::BitfieldDistribution(BitfieldDistributionMessage::NetworkBridgeUpdate(n))
+}
+
+/// The bitfield distribution subsystem.
+pub struct BitfieldDistribution;
+
+impl BitfieldDistribution {
+	/// The protocol identifier for bitfield distribution.
+	const PROTOCOL_ID: ProtocolId = *b"bitd";
+
+	/// Start processing work as passed on from the Overseer.
+	async fn run<Context>(mut ctx: Context) -> SubsystemResult<()>
+	where
+		Context: SubsystemContext<Message = BitfieldDistributionMessage>,
+	{
+		// startup: register the network protocol with the bridge.
+		ctx.send_message(AllMessages::NetworkBridge(
+			NetworkBridgeMessage::RegisterEventProducer(Self::PROTOCOL_ID, network_update_message),
+		))
+		.await?;
+
+		// work: process incoming messages from the overseer and process accordingly.
+		let mut tracker = Tracker::default();
+		loop {
+			{
+				let message = ctx.recv().await?;
+				match message {
+					FromOverseer::Communication { msg } => {
+						// another subsystem created this signed availability bitfield messages
+						match msg {
+							// relay_message a bitfield via gossip to other validators
+							BitfieldDistributionMessage::DistributeBitfield(
+								hash,
+								signed_availability,
+							) => {
+								trace!(target: "bitd", "Processing DistributeBitfield");
+								let job_data = &mut tracker.per_relay_parent.get_mut(&hash).expect("Overseer does not send work items related to relay parents that are not part of our workset. qed");
+
+								let validator = {
+									job_data
+										.validator_set
+										.get(signed_availability.validator_index() as usize)
+										.expect("Our own validation index exists. qed")
```

Comment:
Yeah, again I don't find this to be a convincing proof. It's better to avoid cross-module and cross-crate assumptions for these panic-proofs, because there is nothing stopping a subsystem from sending us a bad piece of data.

Reply:
Garbage-in-garbage-out, but better not to take the node down.